src/System.Private.CoreLib/shared/System/Text/Unicode/Utf8Utility.Validation.cs

```diff
@@ -281,6 +284,96 @@ private static void _ValidateAdditionalNIntDefinitions()
                         pInputBuffer += 4;
                         tempUtf16CodeUnitCountAdjustment -= 2; // 4 UTF-8 code units -> 2 UTF-16 code units (and 2 scalars)
 
+                        if (Sse2.IsSupported && Bmi1.IsSupported)
+                        {
+                            // Optimization: If we see 2-byte subsequences in the text, perhaps the text consists
```

Comment:
How does this apply to emoji?

Reply:
It doesn't - emoji would be detected as a 4-byte sequence and would exit the loop. (It would cause the `(x & 0x7FFF) == 0` check to fail.)